File: src/main/java/com/tambapps/fft4j/fourier/fft_1d/CooleyTukeyRecursiveFFT.java
package com.tambapps.fft4j.fourier.fft_1d;

import com.tambapps.fft4j.complex.Complex;
import com.tambapps.fft4j.complex.vector.ArrayCVector;
import com.tambapps.fft4j.complex.vector.CVector;

/**
 * Compute the FFT in the given vector
 * with the recursive Cooley-Tukey algorithm
 *
 * @link from https://rosettacode.org/wiki/Fast_Fourier_transform
 */
public class CooleyTukeyRecursiveFFT implements FastFourierTransform {

  @Override
  public void compute(CVector vector) {
    computeCopy(vector).copy(vector);
  }

  @Override
  public CVector computeCopy(CVector vector) {
    int N = vector.getSize();
    if (N <= 1) {
      return vector;
    }
    CVector evens = computeCopy(evensCopy(vector));
    CVector odds = computeCopy(oddsCopy(vector));

    Complex[] T = new Complex[N / 2];
    for (int i = 0; i < N / 2; i++) {
      T[i] = odds.getAt(i).multiply(Complex.expI(-2d * Math.PI * ((double) i) / ((double) N)));
    }

    CVector result = new ArrayCVector(N);
    for (int i = 0; i < N / 2; i++) {
      result.setAt(i, evens.getAt(i).plus(T[i]));
      result.setAt(i + N / 2, evens.getAt(i).minus(T[i]));
    }
    return result;
  }

  private CVector evensCopy(CVector vector) {
    int size = (vector.getSize() + 1) / 2;

    CVector copy = new ArrayCVector(size);
    int count = 0;
    for (int i = 0; i < vector.getSize(); i += 2) {
      copy.setAt(count, vector.getAt(i));
      count++;
    }
    return copy;
  }

  private CVector oddsCopy(CVector vector) {
    int size = vector.getSize() / 2;

    CVector copy = new ArrayCVector(size);
    int count = 0;
    for (int i = 1; i < vector.getSize(); i += 2) {
      copy.setAt(count, vector.getAt(i));
      count++;
    }
    return copy;
  }

  @Override
  public String getName() {
    return "Cooley-Tukey recursive";
  }

  @Override
  public String getDescription() {
    return "Cooley-Tukey algorithm implemented recursively (input array must have power of 2 sizes)";
  }
}

File: src/main/kotlin/com/tambapps/fft4j/fourier/fft_1d/CooleyTukeyRecursiveFFT.kt
package com.tambapps.fft4j.fourier.fft_1d

import com.tambapps.fft4j.complex.Complex
import com.tambapps.fft4j.complex.vector.ArrayCVector
import com.tambapps.fft4j.complex.vector.CVector

class CooleyTukeyRecursiveFFT: AbstractFastFourierTransform() {

    override val name = "Cooley-Tukey recursive"
    override val description = "Cooley-Tukey algorithm implemented recursively (input array must have power of 2 sizes)"

    override fun compute(vector: CVector): CVector {
        val N: Int = vector.size
        if (N <= 1) {
            return vector
        }
        val evens = compute(evensCopy(vector))
        val odds = compute(oddsCopy(vector))

        val T = Array(N / 2) { i -> odds[i] * Complex.expI(-2.0 * Math.PI * i.toDouble() / N.toDouble()) }

        val result: CVector = ArrayCVector(N)
        for (i in 0 until N / 2) {
            result[i] = evens[i] + T[i]
            result[i + N / 2] = evens[i] - T[i]
        }
        return result
    }


    private fun evensCopy(vector: CVector): CVector {
        val size: Int = (vector.size + 1) / 2
        return copy(vector, size, 0)
    }

    private fun oddsCopy(vector: CVector): CVector {
        val size: Int = vector.size / 2
        return copy(vector, size, 1)
    }
    
    private fun copy(vector: CVector, size: Int, iStart: Int): CVector {
        val copy: CVector = ArrayCVector(size)
        var count = 0
        var i = iStart
        while (i < vector.size) {
            copy[count] = vector[i]
            count++
            i += 2
        }
        return copy
    }
}